onelogin_rules/onelogin_active_login_activity.py

```diff
@@ -0,0 +1,36 @@
+import time
+from panther_oss_helpers import get_string_set, put_string_set, set_key_expiration, add_to_string_set  # pylint: disable=import-error,line-too-long
+
+THRESH = 2
+THRESH_TTL = 43200  # 1/2 day
+
+
+def rule(event):
+    # Pre-filter: event_type_id = 5 is login events.
+    if event.get('event_type_id') != 5 or not event.get(
+            'ipaddr') or not event.get('user_id'):
+        return False
+    # This track multiple successful logins for different accounts from the same ip address
+    # First, keep a list of unique usernames that have logged in from this ip address
+    event_key = get_key(event)
+    usernames = get_string_set(event_key)
+    if not usernames:
+        # store this as the first user login from this ip address
+        put_string_set(event_key, [event.get('user_id')])
```

Comment:
I have several threshold/key checking rules. Is there a formal way to test these? I'm unsure if I used all of them properly.

Reply:
Unfortunately there is not currently a great way to test rules or policies that rely on API calls outside of the python engine (which is what `put_string_set` and `get_string_set` are doing under the hood). What we can do is try to test the logic leading up to these points for any branches that exit early before making the API call.